ambry-router/src/main/java/com.github.ambry.router/PutOperation.java

```diff
@@ -455,11 +457,59 @@ void clear() {
       chunkIndex = -1;
       chunkBlobId = null;
       chunkException = null;
-      state = ChunkState.Free;
       failedAttempts = 0;
       partitionId = null;
-      correlationIdToChunkPutRequestInfo.clear();
       attemptedPartitionIds.clear();
+      maybeUpdateDefunctBufferInfos();
+      correlationIdToChunkPutRequestInfo.clear();
+      // this assignment should be the last statement as this immediately makes this chunk available to the
+      // ChunkFiller thread for filling.
+      state = ChunkState.Free;
+    }
+
+    /**
+     * Go through the list of requests for which responses were not received, and if there are any that are not yet
+     * sent out completely, add the associated buffer to the defunct list for freeing in the future.
+     */
+    private void maybeUpdateDefunctBufferInfos() {
+      ArrayList<PutRequest> requestsAwaitingSendCompletion = null;
+      for (Map.Entry<Integer, ChunkPutRequestInfo> entry : correlationIdToChunkPutRequestInfo.entrySet()) {
+        if (!entry.getValue().putRequest.isSendComplete()) {
+          if (requestsAwaitingSendCompletion == null) {
+            requestsAwaitingSendCompletion = new ArrayList<>();
+          }
+          requestsAwaitingSendCompletion.add(entry.getValue().putRequest);
+        }
+      }
+
+      if (requestsAwaitingSendCompletion != null) {
+        // This means that the buffer associated with this PutChunk could get read by the NetworkClient in the
+        // future and assigning this PutChunk to a subsequent chunk of the overall blob could lead to this buffer
+        // getting read and written concurrently, or other undefined behavior. There are multiple ways to handle this,
+        // and the simplest way is to set the buf to null so that it gets allocated afresh if/when this PutChunk gets
+        // assigned for a subsequent chunk of the overall blob. Every time this chunk gets polled, an attempt to clear
+        // out the list will be made.
+        defunctBufferInfos.add(new DefunctBufferInfo(buf, requestsAwaitingSendCompletion));
+        buf = null;
```

Comment:
can also allocate new one right here. We are not doing that because it might be wasted?